### PR TITLE
Nested tuple

### PR DESCRIPTION
**Pull request by bluejekyll**
this builds off #46 

I think it's correct :) I need to find some other validations from one of the other binging tests...

### CODE REVIEW

File: foundationdb/src/tuple.rs [hunk 1]
```diff
@@ -216,6 +216,37 @@ impl Single for String {
...
+    fn decode(buf: &[u8]) -> Result<(Self, usize)> {
+        if buf.len() < 2 {
+            return Err(TupleError::EOF);
+        }
+
+        NESTED.expect(buf[0])?;
+
+        let mut tuples = Vec::new();
+        let mut idx = 1;
+        while buf[idx] != EMPTY {
```
Comment (yjh0502) [inline foundationdb/src/tuple.rs]: We should check index boundary condition. Following code panics with out-of-bound

```
SingleValue::decode(&[NESTED, INTZERO])
```

**Comment by yjh0502** [comment]
By the way I rebased and force-pushed the branch to fix conflict before review.

**Comment by bluejekyll** [comment]
This should guard against the out-of-bounds issue. I also changed the EMPTY type to NIL...

At some future point, we may want to split decode and encode into different traits so that we can more generically implement encode, where we can't with decode.

**Comment by bluejekyll** [comment]
let's not merge this yet. I want to see if there's another option to better deal with TupleValue...

**Pull request merged**